### DISCUSSION TITLE
Add a public key on alias to let our friend use it to encrypt mail between simple login and them.

### DISCUSSION THREAD

**Discussion by ghost** [simple-login/app#689] [Feature Requests]
Like the title say.
It's a way to encrypt incomming email from our friend to your services.
With this feature coupled to 

https://github.com/simple-login/app/discussions/687 and https://github.com/simple-login/app/discussions/429

We can have a totally encrypted communication with simple login in the middle.

**Comment by nguyenkims** [comment]
It's actually already possible with an external PGP software: you can generate a key pair for your alias and send the public key to friends. When an email come in, SimpleLogin forwards them as-is and you can decrypt the email with your private key.

This way SimpleLogin isn't aware of the private/public key and it's better for users privacy.

**Reply by ghost** [reply]
yes i know but i mean there i would like someting like an unique public pgp key for your server who can decrypt all incoming pgp mail.
Like that if i generate for instance blabla@simplelogin.test and pass the public key of your server to for exemple (only one who come to my mind) facebook with their pgp encrypted email). their communication with your server will be pgp encrypted, decrypted by the server and forwarded to our inbox (encrypted or not depend if we have configured pgp on our mailbox).

Edit : this feature request is not to prevent letting your server eventually read the content (i tell eventually since it's possible even if it's propably (certainly) not the case) but more protect mail from MITM or other interception BEFORE reaching you.

**Reply by ghost** [reply]
Note : it can allow other thing like external service who detect simplelogin alias to autmatically encrypt mail to you with pgp if you have one unique pgp key for all allias incoming.

**Reply by nguyenkims** [reply]
We have tried to set a common public key for all aliases but unfortunately PGP doesn't support wildcard address, i.e. we can't declare that *@aleeas.com all have a common PGP key.

**Reply by Satlinker** [reply]
I still can't manage to install an end-to-end encryption. Maybe I am doing something wrong: when I want to encrypt a mail with openPGP in thunderbird and send it to a xyz@aleeas.com address I get the error message:
The message cannot be sent with end-to-end encryption because there are problems with the keys of the following recipients: xyz@aleeas.com
OpenPGP requires a public key from the aleeas address. It obviously cannot simply forward it?

**Reply by nguyenkims** [reply]
Have you generated a PGP key for xyz@aleeas.com and import it within Thunderbird? I don't use Thunderbird but as an email alias is just an email address, it should normally work. For example, hi@simplelogin.io is an alias and has a PGP key attached to it.

**Reply by alita-moore** [reply]
I don't know very much, so take this with a grain of salt.

On my thunderbird client it complains about the sender not being the owner of the key. Although the encryption is presumably still okay. I think it's just (rightfully) concerned about the dead man in the middle. But is there a way to just convince the client that it's from the sender even though the sender is not the messenger?

Seems like a catch 22 and I thought maybe I could host simple login locally so I can hold new keys for every alias which would decrypt incoming and re-encrypt back to the sender.

**Comment by ghost** [comment]
> We have tried to set a common public key for all aliases but unfortunately PGP doesn't support wildcard address, i.e. we can't declare that *@aleeas.com all have a common PGP key.

What about generate the key on the browser and upload it to the server ?
yes one key by alias (and must be done manually).

Because i have somthing like 30 or 40 alias, and on them if a had the possibility only 3 or 4 will have a pgp entry (since service who propose pgp encrypted email is not that common, so only my personnal email will have it). maybe limit the number of alias who can have it (at first) to see if it's not too heavy for the server and authorise more if not.

**Reply by nguyenkims** [reply]
Though this can work, there are better softwares out there to generate a PGP key with a lot of options so user can choose. I think for people used to work with PGP, they would maybe prefer having the control in this process and to be sure that the private key never leaves their computer.

**Reply by ghost** [reply]
> Though this can work, there are better softwares out there to generate a PGP key with a lot of options so user can choose. I think for people used to work with PGP, they would maybe prefer having the control in this process and to be sure that the private key never leaves their computer.

it's true but the problem is "IOS" they don't have proper mail client with pgp support so we must rely on services like protonmail and so your idea (who is totally valid don't work for us) :/ .

**Comment by ghost** [comment]
@nguyenkims After some try, i have found out a major problem with using our own pgp key, we cannot use services like protonmail or so, because we must in this case link our pgp key from them (who also leak our email address) so do it force us to use incredibly unconvegnient local software or destroy the major feature of your service (hide our email) and import a key (in their UI) who isn't controlled by them is a big nope (because it just refuse our key).

**Reply by nguyenkims** [reply]
ProtonMail only creates a key for your email indeed. To generate another key, a PGP software is needed. There are several options here, for example the gpg command (https://guides.library.illinois.edu/data_encryption/gpgcheatsheet), [GPGTools](https://gpgtools.org).

**Reply by ghost** [reply]
True but you are bounded to use external software since importing gpg from an gpgtool (like gpg4win) into proton is impossible (refuse it), and mailvelope don't work since they take the encrypted email as file we must download so it's nearly unusable on popular webmail (protonmail, tutanota (even tried infomaniak's free mail).

**Reply by nguyenkims** [reply]
Have you tried with Thunderbird or GPGTools if you're on Mac? It seems that these softwares are more customisable when dealing with PGP keys.

**Reply by ghost** [reply]
> Have you tried with Thunderbird or GPGTools if you're on Mac? It seems that these softwares are more customisable when dealing with PGP keys.

it's totally true, but actually it is a problem for majority of user who use webased software and or mobile ios who isn't easy to use these pgp key.

**Reply by elamast** [reply]
I'm envisioning something like onion routing for email.  The actual email is locally generated for the intended recipient (it could be pgp-encrypted as well) using an MUA to a localhost relay that wraps the entire message (including the metadata) inside a PGP message to a special recipient on the SLmail server using the SLmail public key.  The SLmail server removes that outer layer and routes the message. If the recipient has pgp set up with a public key on the SLmail server, it is delivered to the recipient encrypted. 

Webmail presents a different challenge since the PGP wrapping would need to be done with something like FlowCrypt on Gmail, Startmail, Protonmail, etc (which all need the public key of the special SLmail recipient).  In this case, the entire from/to/subject and message would need to be formatted to a template in the body of the message that is then PGP-encrypted for and sent to the special recipient on the SLMail server. The server would decrypt the message, parse the body into an actual email message, and route as appropriate.

SPAM in both scenarios is controlled by authorized senders (and aliases).  Only one PGP key is needed for the special recipient mailbox on SLmail. (e.g. pgpbox@aleeas.com) that would strip off the outer PGP layer.

The problem with relying on TLS alone is that it only handles transport encryption (which may or may not be broken) and every server in the mail path has access to the metadata. Obscuring the intended recipient right from the outset seems like a good security measure to make it more difficult for determining who is communicating with whom.

**Comment by alita-moore** [comment]
wouldn't this still be possible if the simple login server decrypted then re-encrypted the contents? There's obv a lot of security and privacy concerns about that but I'd be willing to run a local instance for this functionality... or even buy special hardware

i.e. protonmail -> simplelogin (gives pgp key) -> alias (has own pgp pair) -> recipient receives encrypted message with alias' key

**Reply by ghost** [reply]
i'm for it but it was rejected many time already ;(

**Reply by alita-moore** [reply]
Why was it rejected?

**Reply by D3an1el** [reply]
Hi Curious is this going to be a future feature? point less if you can add the contacts pub pgp key and the alias doesn't have one or you cant add one to the alias.

**Reply by nguyenkims** [reply]
Can this feature request correspond to what you need? https://github.com/simple-login/app/discussions/467
